pkg/api-server/dataplane_layout_endpoint.go

```diff
@@ -0,0 +1,186 @@
+	"github.com/kumahq/kuma/pkg/core/resources/manager"
+	core_model "github.com/kumahq/kuma/pkg/core/resources/model"
+	"github.com/kumahq/kuma/pkg/core/resources/store"
+	rest_errors "github.com/kumahq/kuma/pkg/core/rest/errors"
+	"github.com/kumahq/kuma/pkg/core/user"
+	util_slices "github.com/kumahq/kuma/pkg/util/slices"
+	xds_context "github.com/kumahq/kuma/pkg/xds/context"
+)
+
+type dataplaneLayoutEndpoint struct {
+	resManager         manager.ResourceManager
+	meshContextBuilder xds_context.MeshContextBuilder
+	resourceAccess     access.ResourceAccess
+}
+
+func newDataplaneLayoutEndpoint(resManager manager.ResourceManager, meshContextBuilder xds_context.MeshContextBuilder, resourceAccess access.ResourceAccess) *dataplaneLayoutEndpoint {
+	return &dataplaneLayoutEndpoint{
+		resManager:         resManager,
+		meshContextBuilder: meshContextBuilder,
+		resourceAccess:     resourceAccess,
+	}
+}
+
+func (dle *dataplaneLayoutEndpoint) addEndpoint(ws *restful.WebService) {
+	ws.Route(
+		ws.GET("/meshes/{mesh}/dataplanes/{name}/_layout").
+			To(dle.getLayout).
+			Doc("Get Dataplane Layout").
+			Returns(http.StatusOK, "OK", nil),
+	)
+}
+
+func (dle *dataplaneLayoutEndpoint) getLayout(request *restful.Request, response *restful.Response) {
+	meshName := request.PathParameter("mesh")
+	dataplaneName := request.PathParameter("name")
+
+	if err := dle.resourceAccess.ValidateGet(
+		request.Request.Context(),
+		core_model.ResourceKey{Mesh: meshName, Name: dataplaneName},
+		core_mesh.DataplaneResourceTypeDescriptor,
+		user.FromCtx(request.Request.Context()),
+	); err != nil {
+		rest_errors.HandleError(request.Request.Context(), response, err, "Access Denied")
+		return
+	}
+
+	mesh, err := dle.getMeshResource(request.Request.Context(), meshName)
+	if err != nil {
+		rest_errors.HandleError(request.Request.Context(), response, err, "Failed to retrieve Mesh")
+		return
+	}
+
+	if mesh.Spec.GetMeshServices().GetMode() == v1alpha1.Mesh_MeshServices_Disabled {
+		response.WriteHeader(http.StatusBadRequest)
+		return
+	}
+
+	dataplane, err := dle.getDataplaneResource(request.Request.Context(), meshName, dataplaneName)
+	if err != nil {
+		rest_errors.HandleError(request.Request.Context(), response, err, "Failed to retrieve Dataplane")
+		return
+	}
+
+	baseMeshContext, err := dle.meshContextBuilder.BuildBaseMeshContextIfChanged(request.Request.Context(), meshName, nil)
+	if err != nil {
+		rest_errors.HandleError(request.Request.Context(), response, err, "Failed to build MeshContext")
+	}
+
+	inbounds := util_slices.Map(dataplane.Spec.GetNetworking().GetInbound(), func(inbound *v1alpha1.Dataplane_Networking_Inbound) api_common.DataplaneInbound {
+		sectionName := inbound.GetName()
+		if sectionName == "" {
+			sectionName = strconv.Itoa(int(inbound.GetPort()))
+		}
+
+		return api_common.DataplaneInbound{
+			Kri:      kri.From(dataplane, sectionName).String(),
+			Port:     int32(inbound.GetPort()),
+			Protocol: inbound.GetProtocol(),
```

Comment:
are you going to backfill with the recent addition later?

Reply:
yes, we need to have common logic that will build this new names so we can reuse it in layout endpoint